rclcpp/src/rclcpp/qos.cpp

```diff
@@ -193,6 +193,21 @@ bool operator==(const rmw_time_t & left, const rmw_time_t & right)
 }
 }  // unnamed namespace
 
+bool QoS::compatibility_policies_equal(const QoS & other) const
+{
+  // Note that these policies do not affect QoS compatibility
+  // * History + Depth (relevant only for local behavior)
+  // * Lifespan (only relevant locally on Publishers)
+  // * avoid ros namespace conventions (this affects choice of topic names, not quality of service)
+  const auto & pl = get_rmw_qos_profile();
+  const auto & pr = other.get_rmw_qos_profile();
+  return pl.reliability == pr.reliability &&
+         pl.durability == pr.durability &&
```

Comment:
It's not that simple is it?

For instance, see:

https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/#qos-compatibilities

For instance with durability, if a publisher is `volatile` and a subscribing node asks for `transient local`, they will not communicate, but if the publisher is `transient local` and the subscribing node asks for `volatile`, then they will talk, just as `volatile` instead.

There's a similar asymmetry with reliability.

It seems like this check is insufficient atm, and further more you need to know in what context the QoS will be used, i.e. for either a publisher or subscription.

Reply:
In the comment about this function I address exactly this concern.

This is _not_ a compatibility check. This is a check that all fields that _affect_ compatibility are _exactly equal_

Reply:
In a sense, it is checking a very limited (but bidirectional) subset of compatibility.

Reply:
In rosbag2, we are currently giving up if there are publishers offering different qos policies, we don't have time to make it smarter than that for Foxy. Most ROS systems will have a limited number of publishers on any given topic, and it's extremely likely that all of them will be offering the same QoS policies (at least within the fields that affect compatibility).

Reply:
Well, I guess I'm just dubious of the value of having such a check in the rclcpp API, rather than in rosbag2, where those assumptions hold...

Reply:
We've been talking about this and essentially thought it might be well placed here for maintenance purposes. Meaning, if at some point the QoS have different attributes to compare, we'd only have to modify it in one central place. 
But I am also okay with having this check in rosbag2 if it's really only that tool which would profit from this function.